### PR TITLE
[tcp-ping-pong] Refactor to add Drop

### LINKED ISSUE #719: [tests] Refactor `tcp-ping-pong` to Follow "Associated Functions" Design
## Context

The `tcp-ping-pong` application was written as a collection of standalone functions. 

We should refactor the `tcp-ping-pong` application to follow an "associated functions" code structure, so that we can have correct trait implementations, such as `Drop`.

### PR DESCRIPTION
**Pull request by anandbonde**
Closes #719

### CODE REVIEW

File: examples/rust/tcp-ping-pong.rs [hunk 1]
```diff
@@ -198,178 +197,174 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
...
-                anyhow::bail!("push and wait failed: {:?}", e);
-            }
-            if let Err(e) = libos.sgafree(sga) {
-                // If error, free socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("failed to release scatter-gather array: {:?}", e)
-            }
+    fn run(&mut self, remote: SocketAddrV4, nrounds: usize) -> Result<()> {
+        if let Err(e) = connect_and_wait(&mut self.libos, self.sockqd, remote) {
+            close(&mut self.libos, self.sockqd);
```
Comment (iyzhang) [inline examples/rust/tcp-ping-pong.rs:315]: This close is not necessary, I think.

File: examples/rust/tcp-ping-pong.rs [hunk 2]
```diff
@@ -198,178 +197,174 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
...
 
-            for x in &recvbuf {
-                if *x != fill_char {
-                    close(&mut libos, sockqd);
-                    anyhow::bail!("fill check failed: expected={:?} received={:?}", fill_char, *x);
+            // Pop data, and sanity check it.
+            {
+                let mut recvbuf: [u8; BUFFER_SIZE] = [0; BUFFER_SIZE];
+                if let Err(e) = pop_and_wait(&mut self.libos, accepted_qd, &mut recvbuf) {
+                    close(&mut self.libos, accepted_qd);
```
Comment (iyzhang) [inline examples/rust/tcp-ping-pong.rs:246]: You could add `accepted_qd: Option<QDesc>` to the server data structure and then close it in the drop if there is qd. That way we won't have to specifically call close each time here and on line 252.

**Review by iyzhang — CHANGES_REQUESTED** [review]

File: examples/rust/tcp-ping-pong.rs (hunk 1 above)
Comment (anandbonde) [reply examples/rust/tcp-ping-pong.rs]: Done.

File: examples/rust/tcp-ping-pong.rs (hunk 2 above)
Comment (anandbonde) [reply examples/rust/tcp-ping-pong.rs]: Done.

**Review by iyzhang — APPROVED** [review]

**Pull request merged**